### PR TITLE
Quota VM_DISK tariff calculation

### PR DESCRIPTION
**Pull request by BryanMLima**
### Description

Currently, the quota plugin ignores the tariff calculation for the types below:

* VM_DISK_IO_READ
* VM_DISK_IO_WRITE
* VM_DISK_BYTES_READ
* VM_DISK_BYTES_WRITE

The code was changed as to now calculate the quota tariff for these types. Moreover, the unit types were changed as well, as shown in the table below. It is noteworthy that the raw values for these types are not modified; therefore, the calculation of the tariff for them are: `TARIFF * CONSUMPTION`.

| Type | Unit |
| ------ | ------ |
| VM_DISK_IO_READ | IOPS |
| VM_DISK_IO_WRITE | IOPS |
| VM_DISK_BYTES_READ | Bytes |
| VM_DISK_BYTES_WRITE | Bytes |

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [x] Minor
- [ ] Trivial

### How Has This Been Tested?

This was tested in a local lab creating a quota tariff for each type. Then, I reinstalled a previously created VM and generated a quota statement.

Creating a quota tariff with type `VM_DISK_IO_READ`:
~~~
(lab-bryan) 🐱 > quota tariffcreate name=mr-443-21 usagetype=21 value=30
{
  "quotatariff": {
    "currency": "$",
    "effectiveDate": "2023-01-24T15:39:39+0000",
    "name": "mr-443-21",
    "tariffValue": 30,
    "usageDiscriminator": "None",
    "usageName": "VM_DISK_IO_READ",
    "usageType": 21,
    "usageTypeDescription": "VM Disk usage(I/O Read)",
    "usageUnit": "IOPS",
    "uuid": "e580d696-5e5e-4423-87e5-130298a2cb0f"
  }
}
~~~

~~~
(lab) 🐱 > quota statement accountid=af16aaed-26de-11ec-8dcf-5254005dcdac account=admin domainid=52d83793-26de-11ec-8dcf-5254005dcdac startdate=2023-01-23 enddate=2023-01-25
{
  "statement": {
    "currency": "$",
    "enddate": "2023-01-25T23:59:59+0000",
    "quotausage": [
      (...)
      {
        "accountid": "af16aaed-26de-11ec-8dcf-5254005dcdac",
        "domain": "52d83793-26de-11ec-8dcf-5254005dcdac",
        "name": "VM_DISK_IO_READ",
        "quota": 771439.82,
        "type": 21,
        "unit": "IOPS"
      },
      {
        "accountid": "af16aaed-26de-11ec-8dcf-5254005dcdac",
        "domain": "52d83793-26de-11ec-8dcf-5254005dcdac",
        "name": "VM_DISK_IO_WRITE",
        "quota": 0.44,
        "type": 22,
        "unit": "IOPS"
      },
      {
        "accountid": "af16aaed-26de-11ec-8dcf-5254005dcdac",
        "domain": "52d83793-26de-11ec-8dcf-5254005dcdac",
        "name": "VM_DISK_BYTES_READ",
        "quota": 12667348.48,
        "type": 23,
        "unit": "Bytes"
      },
      {
        "accountid": "af16aaed-26de-11ec-8dcf-5254005dcdac",
        "domain": "52d83793-26de-11ec-8dcf-5254005dcdac",
        "name": "VM_DISK_BYTES_WRITE",
        "quota": 2929711.64,
        "type": 24,
        "unit": "Bytes"
      }
      (...)
    ],
    "startdate": "2023-01-23T00:00:00+0000",
    "totalquota": 16368503.29
  }
}
~~~

### CODE REVIEW

**Review by DaanHoogland — APPROVED** [review]
clgtm,
Is there any compelling reason to have this included in 4.18, @BryanMLima @GutoVeronezi , as the custom quota tarrifs is also part of 4.18?

**Comment by GutoVeronezi** [comment]
> clgtm, Is there any compelling reason to have this included in 4.18, @BryanMLima @GutoVeronezi , as the custom quota tarrifs is also part of 4.18?

@DaanHoogland, previously to introducing the custom quota tariffs, those usage types already were not calculated:

- code in 4.17.2.0:
https://github.com/apache/cloudstack/blob/5b9a989ab0dd6b8f482f4d6c6ce24e1a6108002e/framework/quota/src/main/java/org/apache/cloudstack/quota/QuotaManagerImpl.java#L185-L190

It would be good to have this in 4.18; however, I do not see it as mandatory. In the worst case, nothing changes to the final user.

**Review by rohityadavcloud — APPROVED** [review]
Didn't test it but changes LGTM.

File: engine/schema/src/main/resources/META-INF/db/schema-41720to41800.sql [hunk 1]
```diff
@@ -1079,3 +1079,7 @@ CREATE TABLE IF NOT EXISTS `cloud`.`console_session` (
 INSERT INTO `cloud`.`role_permissions` (`uuid`, `role_id`, `rule`, `permission`) VALUES (UUID(), 2, 'assignVolume', 'ALLOW');
 INSERT INTO `cloud`.`role_permissions` (`uuid`, `role_id`, `rule`, `permission`) VALUES (UUID(), 3, 'assignVolume', 'ALLOW');
 
+
+-- Increases the overall size of the decimal places of the column `quota_used` from 15 to 20.
```
Comment (GutoVeronezi) [inline engine/schema/src/main/resources/META-INF/db/schema-41720to41800.sql:1083]:
```suggestion
-- Increases the precision of the column `quota_used` from 15 to 20, keeping the scale of 8.
```

**Review by GutoVeronezi — APPROVED** [review]
CLGTM

@BryanMLima, I pointed out a minor adjust in the comment in SQL.

**Review by stephankruggg — APPROVED** [review]
CLGTM

**Comment by DaanHoogland** [comment]
@GutoVeronezi cc @BryanMLima , I am happy to include this in 4.18, especially as the unit test is adjusted to include this and the risk for other features is minimal. Is there any more automated testing that could be added for this?

**Comment by BryanMLima** [comment]
@DaanHoogland cc @GutoVeronezi Thanks for adding to version `4.18`. Regarding the automated testing, I don't think so, as it is a rather simple addition.

**Pull request merged**